IMAGING-266 read numeric data from GeoTIFFs

gwlucastrig opened the pull request:
Updates to the TIFF API to support access to numeric data from TIFF files.  This change addresses JIRA item IMAGING-266.

kinow reviewed (commented):
Guess we will have to either update the code, or add exclusion for Spotbugs/Findbugs @gwlucastrig : https://github.com/apache/commons-imaging/pull/165/checks?check_run_id=3551411645

gwlucastrig commented:
Sorry, I forgot about the Spotbugs test (did checkstyle, pmd, and findbugs, but not Spotbugs).  So I just submitted changes for Spotbugs.  

For reasons of efficiency, the TiffRasterData classes expose arrays directly through the API. This approach is reasonable because the raster data classes are really just simple containers for transferring image read results to the calling application.  Also, since many TIFF rasters consist of literally 100s of millions of raster cells, we wish to avoid making redundant copies of data in memory.  This rationale is described in the Javadoc for the relevant methods with appropriate caveats for developers.

kinow commented on src/test/java/org/apache/commons/imaging/formats/tiff/TiffRasterDataIntTest.java:
```diff
@@ -0,0 +1,203 @@
+            fail("Constructor did not detect bad width");
+        }catch(final IllegalArgumentException illArgEx){
+            // success!
+        }
```
@gwlucastrig I believe the tests that are using this pattern `try { something; fail(); } catch {}` is JUnit 4. But for Junit 5 we can use:

```java
assertThrows(IllegalArgumentException.class, () -> new TiffRasterDataInt(-1, 10));
``` 

There are other places in this PR where this can be applied, you can look at GitHub's UI to search for this "fail(" text if you'd like.

gwlucastrig replied:
done

kinow commented on src/test/java/org/apache/commons/imaging/formats/tiff/TiffRasterDataIntTest.java:
```diff
@@ -0,0 +1,203 @@
+			assertEquals((int)result[i], data[i]);
+		}
+		final int []iResult = raster.getIntData();
+        assertArrayEquals(data, iResult);
```
Tabs and spaces here @gwlucastrig :point_up: (better confirm other files don't have tabs, only spotted it here due to the lines not aligned due to mix spaces/tabs, but we could have in the other files).

gwlucastrig replied:
will look for these.

kinow reviewed (changes requested):
@gwlucastrig calling it a day, but managed to review 11 or 16 files :+1: Take a look at my comments, but feel free to decide what must be updated or not. Thanks!!!

gwlucastrig commented:
Found a bug in TiffRasterInt, getDataType returns FLOAT instead of INTEGER.  I will fix it when I address the other review comments.

gwlucastrig commented on src/test/java/org/apache/commons/imaging/formats/tiff/TiffShortIntRoundTripTest.java:
```diff
@@ -0,0 +1,229 @@
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import java.awt.image.BufferedImage;
```
added a new test that uses it.

gwlucastrig commented on src/main/java/org/apache/commons/imaging/formats/tiff/TiffImageParser.java:
```diff
@@ -762,12 +763,12 @@ public void writeImage(final BufferedImage src, final OutputStream os, final Map
-     * following code illustrates the approach:
+     * If desired, sub-image data can be read from the file by using a Java
+     * {@code Map} instance to specify the subsection of the image that
+     * isrequired. The following code illustrates the approach:
```
done

gwlucastrig commented on src/test/java/org/apache/commons/imaging/formats/tiff/TiffRasterDataIntTest.java:
```diff
@@ -0,0 +1,203 @@
+                int test = (int) instance.getValue(x, y);
+                assertEquals(index, test, "Set/get value test failed");
+				instance.setIntValue(x, y, index);
+                test = (int) instance.getIntValue(x, y);
```
done

gwlucastrig commented on src/test/java/org/apache/commons/imaging/formats/tiff/TiffRasterDataIntTest.java:
```diff
@@ -0,0 +1,203 @@
+        int k = 0;
+        for (int i = 0; i < width; i++) {
+            for (int j = 0; j < height; j++) {
+                data[k] = (int)k;
```
done

gwlucastrig commented on src/test/java/org/apache/commons/imaging/formats/tiff/TiffRasterDataIntTest.java:
```diff
@@ -0,0 +1,203 @@
+                final int index = y * width + x;
+                int test = (int) raster.getValue(x, y);
+                assertEquals(index, test, "Get into source data test failed at (" + x + "," + y + ")");
+                test = (int) raster.getIntValue(x, y);
```
done

gwlucastrig commented:
Made changes for your review comments and added a few more JUnit tests to address the Coverage issues.  

Unfortunately, I accidentally picked up two files that had tabs in them when I replaced tabs with spaces.  Other than replacing tabs with spaces, these files did not change.

\test\java\org\apache\commons\imaging\TestImageReadException.java  
\test\java\org\apache\commons\imaging\TestImageWriteException.java

gwlucastrig commented:
I think that's everything...   Thanks for your excellent review.

Github is reporting that more changes are requested, but searching through this discussion, I don't see anything that's unresolved.

Please let me know if you find additional items that need attention.

kinow commented:
Thanks @gwlucastrig !

>Github is reporting that more changes are requested, but searching through this discussion, I don't see anything that's unresolved.

It's because when I submitted my review, I had marked it as request changes. It will remain with that status until I review it again. We are in lockdown in Auckland due to the delta variant, so quite sure I'll be able to finish reviewing it over this weekend.

gwlucastrig commented:
Hope you and your wife stay healthy.

On Fri, Sep 10, 2021, 5:38 PM Bruno P. Kinoshita ***@***.***>
wrote:

> Thanks @gwlucastrig <https://github.com/gwlucastrig> !
>
> Github is reporting that more changes are requested, but searching through
> this discussion, I don't see anything that's unresolved.
>
> It's because when I submitted my review, I had marked it as request
> changes. It will remain with that status until I review it again. We are in
> lockdown in Auckland due to the delta variant, so quite sure I'll be able
> to finish reviewing it over this weekend.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/commons-imaging/pull/165#issuecomment-917227883>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEWJDYKARSB6OHMRODCUNKTUBJ3FZANCNFSM5DWCCJSA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

kinow commented on src/main/java/org/apache/commons/imaging/formats/tiff/datareaders/DataReaderStrips.java:
```diff
@@ -390,16 +405,69 @@ public TiffRasterData readRasterData(final Rectangle subImage)
-                decompressed,
-                predictor, bitsPerPixel, byteOrder);
+                    width,
+                    (int) rowsInThisStrip,
```
Unnecessary casting.

gwlucastrig replied:
done

kinow commented on src/main/java/org/apache/commons/imaging/formats/tiff/datareaders/DataReaderStrips.java:
```diff
@@ -390,16 +405,69 @@ public TiffRasterData readRasterData(final Rectangle subImage)
+                    width,
+                    decompressed,
+                    predictor, bitsPerPixel, byteOrder);
             transferBlockToRaster(0, yStrip, width, (int) rowsInThisStrip, blockData,
```
Also unnecessary casting, but not in this PR.

gwlucastrig replied:
Done

kinow commented on src/main/java/org/apache/commons/imaging/formats/tiff/datareaders/DataReaderStrips.java:
```diff
@@ -390,16 +405,69 @@ public TiffRasterData readRasterData(final Rectangle subImage)
+                    bytesPerStrip, width, rowsInThisStrip);
+            final int[] blockData = unpackIntSamples(
+                    width,
+                    (int) rowsInThisStrip,
```
Unnecessary casting.

gwlucastrig replied:
done

kinow commented on src/main/java/org/apache/commons/imaging/formats/tiff/datareaders/DataReaderStrips.java:
```diff
@@ -390,16 +405,69 @@ public TiffRasterData readRasterData(final Rectangle subImage)
+                    width,
+                    decompressed,
+                    predictor, bitsPerPixel, byteOrder);
+            transferBlockToRaster(0, yStrip, width, (int) rowsInThisStrip, blockData,
```
Unnecessary casting.

gwlucastrig replied:
Done

kinow reviewed (commented):
Finished reviewing the rest of the code. Looks good to me @gwlucastrig . Haven't looked at coverage, but from reading the tests looks like you covered most (if not all) of the new code. Left a few comments. Also, do you think you could rebase & squash the commits to have a single commit? Otherwise I can rebase & squash when updating changes.xml and JIRA.

Thanks @gwlucastrig

gwlucastrig commented on src/main/java/org/apache/commons/imaging/formats/tiff/datareaders/DataReaderTiled.java:
```diff
@@ -286,7 +288,20 @@ public ImageBuilder readImageData(final Rectangle subImageSpecification,
-        throws ImageReadException, IOException {
+            throws ImageReadException, IOException {
+        switch (sampleFormat) {
+            case TiffTagConstants.SAMPLE_FORMAT_VALUE_IEEE_FLOATING_POINT:
```
It took a bit of searching to find what this is about.

    public static final int SAMPLE_FORMAT_VALUE_COMPLEX_INTEGER = 5;
    public static final int SAMPLE_FORMAT_VALUE_IEEE_FLOATING_POINT_1 = 6;

Sample formats 5 (complex integer) and 6 (complex float) are not actually part of the TIFF spec, but seem to have been introduced by LibTIFF at some point.  There is no support for complex-conjugate-pairs in Commons Imaging at this time. Neither of these values are used elsewhere in the the code.

At the very least, I will rename FLOATING_POINT_1 to be COMPLEX_FLOAT.

You can read more at

[Extension TIFF Tag Sample Format](https://www.awaresystems.be/imaging/tiff/tifftags/sampleformat.html)

gwlucastrig commented on src/main/java/org/apache/commons/imaging/formats/tiff/datareaders/DataReaderTiled.java:
```diff
@@ -321,19 +336,68 @@ public TiffRasterData readRasterData(final Rectangle subImage)
+        final int col0 = xRaster / tileWidth;
+        final int col1 = (xRaster + rasterWidth - 1) / tileWidth;
+        final int row0 = yRaster / tileLength;
+        final int row1 = (yRaster + rasterHeight - 1) / tileLength;
```
Actually, tile length is what they call it in the TIFFF specification.  Not sure they used that term, though it might be a holdover from the earlier terminology strip-length. 

In either case, I think we should leave this one as is...

gwlucastrig commented on src/main/java/org/apache/commons/imaging/formats/tiff/datareaders/ImageDataReader.java:
```diff
@@ -519,8 +516,55 @@ protected void applyPredictorToBlock(final int width, final int height, final in
         return samples;
     }
 
+    protected int[] unpackIntSamples(
```
done

gwlucastrig commented on src/main/java/org/apache/commons/imaging/formats/tiff/datareaders/ImageDataReader.java:
```diff
@@ -519,8 +516,55 @@ protected void applyPredictorToBlock(final int width, final int height, final in
+        final int predictor,
+        final int bitsPerSample,
+        final ByteOrder byteOrder)
+        throws ImageReadException {
```
done

gwlucastrig commented:
I looked up rebasing and squashing commits and I wanted to check with you before proceeding.

I have 6 commits that I just pushed as part of this PR.  So would my command be the following?

> git rebase -i HEAD~6

But as I was researching this, I found a comment:

> If you have previously pushed your code to a remote branch, you will need to force push.
> 
> git push origin branchName --force

Do I need to do the force when I push?   And, if so, would my command be the following?

> git push origin IMAGING-266a --force

where IMAGING-266a is the branch name for these changes?

kinow commented:
Hi @gwlucastrig ! 

Great job fixing the code comments and trying to rebase.

You have done everything correctly. And you will have indeed to push force.

I always take a look at the remote branch in GitHub, and compare with my local branch (using gitk).

If I'm happy that my local branch is the way I want my remote to look, then I push force.

Note that you should avoid push-force to certain repos in case of existijg etiquette. E.g. linux project forbids push force to their main branches. Apache projects normally discourage too.

But you are pushing force to your branch, to your fork. So it doesn't matter in this case. It's only to make merging the code easier :)

I will take a look at the code latert oday gain, if nothing wrong will include one commit later with changes.xml and merge it! 👏

p.s.: a git rebase squash is an advanced git topic, great job! It is also part of the process for fixing conflicts, so you got most of what's needed to fix conflicts in the future

p.p.s: I forgot to add a comment about scansize. Is it OK if I edit your commit when merging, and rename it to scanSize?

gwlucastrig commented:
I just pushed the squashed version of the code.  

Please go ahead and rename scansize to scanSize.  I definitely like that name better.

kinow commented:
@gwlucastrig great job! GitHUb UI is already showing 1 commit, so you have successfully rebased and squashed your commits :-)

I had marked during the previous review all files reviewed as "Viewed". Now, when I go to Files Changed, @gwlucastrig , GitHub keeps the files that did not change closed. And it expands the ones that suffered modifications since I left reviewed it.

![image](https://user-images.githubusercontent.com/304786/132966030-1a3b14d0-e45a-40eb-8467-0fef30be005b.png)

This saves a lot of time since I don't need to re-read the code to compare with the previous version.

kinow commented:
>I am really looking forward to going onto the JIRA page and posting a notice about our progress on ISSUE 266 (including the picture of Auckland). Commons Imaging is shaping up to be one of the most useful TIFF solutions out there... Certainly, it's my favorite Java solution.

And props go to **you** for improving the API and code so much. Really appreciate all your work in reviewing issues, sending new code (with the extra work of chasing test images), and in educating other devs like myself :)

kinow reviewed (approved):
Changes look good! LGTM

Thank you @gwlucastrig

kinow commented:
@gwlucastrig here's what I've done in the last commit.

1. Renamed `scansize` to `scanSize` in `ImageDataReader.java`
2. `git commit -am 'f'` (just so I know I need to squash it)
3. `git rebase -i HEAD~2` and I replace `pick f` by `f f`
4. Added an entry in `changes.xml`. This file is used to generate the changelog and in the site and in the `RELEASE-NOTES.txt` file (I think, although can't recall if we use JIRA export tool in release process?). I added an entry with myself as dev, and you as due-to. The dev is always an ASF committer, and due-to the person responsible for work (it's not always present). I didn't put your user as dev as you are still not a committer (only pending you accept & submit the form :)
5. Pushed force a commit with the squashed commit and the `changes.xml` (I can push to your branch as you allowed it when creating the PR)
6. Waited for CI to confirm I didn't mess it up, then merged the PR
7. Updated the JIRA (fix version, status, etc).

This is the whole process for a ticket in Imaging, but applies to pretty much every Apache Commons projects too.

Cheers
Bruno
